packages/horizon/contracts/HorizonStaking.sol

```diff
@@ -0,0 +1,869 @@
+            prov.tokens = prov.tokens - providerTokensSlashed;
+            serviceProviders[_serviceProvider].tokensProvisioned =
+                serviceProviders[_serviceProvider].tokensProvisioned -
+                providerTokensSlashed;
+            serviceProviders[_serviceProvider].tokensStaked =
+                serviceProviders[_serviceProvider].tokensStaked -
+                providerTokensSlashed;
+            emit ProvisionSlashed(_serviceProvider, verifier, providerTokensSlashed);
+        }
+
+        tokensToSlash = tokensToSlash - providerTokensSlashed;
+        if (tokensToSlash > 0) {
+            DelegationPoolInternal storage pool;
+            if (verifier == SUBGRAPH_DATA_SERVICE_ADDRESS) {
+                pool = legacyDelegationPools[_serviceProvider];
+            } else {
+                pool = delegationPools[_serviceProvider][verifier];
+            }
+            if (delegationSlashingEnabled) {
+                require(pool.tokens >= tokensToSlash, "insufficient delegated tokens");
+                TokenUtils.burnTokens(_graphToken(), tokensToSlash);
+                uint256 delegationFractionSlashed = (tokensToSlash * FIXED_POINT_PRECISION) / pool.tokens;
+                pool.tokens = pool.tokens - tokensToSlash;
+                pool.tokensThawing =
+                    (pool.tokensThawing * (FIXED_POINT_PRECISION - delegationFractionSlashed)) /
+                    FIXED_POINT_PRECISION;
+                emit DelegationSlashed(_serviceProvider, verifier, tokensToSlash);
+            } else {
+                emit DelegationSlashingSkipped(_serviceProvider, verifier, tokensToSlash);
+            }
+        }
+    }
+
+    /**
+     * @notice Check if an operator is authorized for the caller on a specific verifier / data service.
+     * @param _operator The address to check for auth
+     * @param _serviceProvider The service provider on behalf of whom they're claiming to act
+     * @param _verifier The verifier / data service on which they're claiming to act
+     */
+    function isAuthorized(
+        address _operator,
+        address _serviceProvider,
+        address _verifier
+    ) public view override returns (bool) {
+        if (_operator == _serviceProvider) {
+            return true;
+        }
+        if (_verifier == SUBGRAPH_DATA_SERVICE_ADDRESS) {
+            return legacyOperatorAuth[_serviceProvider][_operator];
+        } else {
+            return operatorAuth[_serviceProvider][_verifier][_operator];
+        }
+    }
+
+    // staked tokens that are currently not provisioned, aka idle stake
+    // `getStake(serviceProvider) - ServiceProvider.tokensProvisioned`
+    function getIdleStake(address serviceProvider) public view override returns (uint256 tokens) {
+        return
+            serviceProviders[serviceProvider].tokensStaked -
+            serviceProviders[serviceProvider].tokensProvisioned -
+            serviceProviders[serviceProvider].__DEPRECATED_tokensLocked;
```

Comment:
```suggestion
        return
            serviceProviders[serviceProvider].tokensStaked -
            serviceProviders[serviceProvider].tokensProvisioned -
            serviceProviders[serviceProvider].__DEPRECATED_tokensLocked - 
            serviceProviders[serviceProvider].__DEPRECATED_tokensAllocated;
```

I think we need to also deduct `tokensAllocated` to prevent service providers from "double dipping" the stake.